### PR TITLE
ON-42104 # Added developer key and app association documentation

### PR DESCRIPTION
**Pull request by mymattcarroll**

### CODE REVIEW

File: DOCUMENTATION.md [hunk 1]
```diff
@@ -51,8 +51,27 @@ Various Types used in this package can also be imported:
...
-Without the associated permissions for a particular action, you will be prohibited from completing that action using the OneBlink SDK. The associated permissions for each function are detailed on the function definitions.
+The majority of actions in this SDK require that your developer key has particular permissions. These permissions can be set in your tenant's console. If you do not have access to your tenant's console, you will need to work with your Administrator to associate the permissions that you require with your key. The associated permissions for each function are detailed on the function definitions.
+
+### Developer Key & Role
+
+Actions that require a developer key with a minimum role permission will have the following in the function definition:
+
+**Minimum Role Permission**
+
+The developer key used must be assigned a role with at least the permission documented for each action. E.g. If action's minimum role permission is _Forms: Read Only_, the role assigned to the developer key could have _Forms: Read Only_ or _Forms: Manager_.
```
Comment (jdawg093) [inline DOCUMENTATION.md:64]:
```suggestion
The developer key used must be assigned a role with at least the permission documented for each action. E.g. If the action's minimum role permission is _Forms: Read Only_, the role assigned to the developer key could have _Forms: Read Only_ or _Forms: Manager_.
```

File: DOCUMENTATION.md [hunk 2]
```diff
@@ -51,8 +51,27 @@ Various Types used in this package can also be imported:
...
+
+### Developer Key & App Association
+
+Actions that require a developer key to be assigned to an app will have the following in the function definition:
+
+**App Association Required**
+
+These actions require the developer key to be associated with an app that is associated with the form that is being actioned. A form can be associated with an app in the following ways:
+
+- For _Forms List_ type apps, the form must be in the Forms List menu item
```
Comment (jdawg093) [inline DOCUMENTATION.md:74]:
```suggestion
- For _Forms List_ type apps, the form must be in the Forms List menu item.
```

File: src/classes/FormElementLists.ts [hunk 3]
```diff
@@ -68,39 +61,35 @@ export default class FormElementLists extends OneBlinkAPI {
...
+   *             label: 'Two',
+   *             value: '2',
+   *           },
+   *         ],
+   *         formsAppEnvironmentId: 1,
+   *       },
+   *     ],
+   *     type: 'STATIC',
+   *   }
+   *   // Use list here...
```
Comment (divporter) [inline src/classes/FormElementLists.ts:89]:
```suggestion
   *   const list = await formElementListsClient.createFormElementList(data)
   *   // Use list here...
```

File: DOCUMENTATION.md [hunk 4]
```diff
@@ -51,8 +51,27 @@ Various Types used in this package can also be imported:
...
+
+The developer key used must be assigned a role with at least the permission documented for each action. E.g. If action's minimum role permission is _Forms: Read Only_, the role assigned to the developer key could have _Forms: Read Only_ or _Forms: Manager_.
+
+### Developer Key & App Association
+
+Actions that require a developer key to be assigned to an app will have the following in the function definition:
+
+**App Association Required**
+
+These actions require the developer key to be associated with an app that is associated with the form that is being actioned. A form can be associated with an app in the following ways:
```
Comment (Zaxist) [inline DOCUMENTATION.md:72]:
```suggestion
In the case of any actions relating to forms, the assigned app must be associated with the form that is being actioned. A form can be associated with an app in the following ways:
```

File: src/classes/sendEmail.ts [hunk 5]
```diff
@@ -5,48 +5,42 @@ import OneBlinkAPI from '../lib/one-blink-api'
 
 import { SendMailOptions, SendEmailResult } from '../types'
 /**
- * ## Example
+ * **Permissions**
  *
- * ```javascript
- * const OneBlink = require('@oneblink/sdk')
+ * In AWS, you will require an IAM Role associated with the email used that has
+ * the permissions in order to use this function:
```
Comment (divporter) [inline src/classes/sendEmail.ts:11]: ❤️

File: DOCUMENTATION.md (hunk 4 above)
Comment (mymattcarroll) [reply DOCUMENTATION.md:72]: This implies that not all of these actions relate to forms which is incorrect. They are all associated with a single form.

File: src/classes/Forms.ts [hunk 6]
```diff
@@ -31,65 +31,48 @@ import {
...
-   * forms.generateFormUrl(parameters).then((result) => {
-   *   const formUrl = result.formUrl
-   *   // Use form URL here...
-   * })
-   * ```
-   *
-   * @param parameters An object containing all parameters to be passed to the
-   *   function
-   * @returns
+   * **App Association Required**
```
Comment (divporter) [inline src/classes/Forms.ts:47]: what about forms that aren't in an app?

File: DOCUMENTATION.md (hunk 2 above)
Comment (Zaxist) [inline DOCUMENTATION.md:74]:
```suggestion
- For _Forms List_ type apps, the form can be in the _Forms List_ menu item.
```

File: DOCUMENTATION.md [hunk 7]
```diff
@@ -51,8 +51,27 @@ Various Types used in this package can also be imported:
...
+
+- For _Forms List_ type apps, the form must be in the Forms List menu item.
```
Comment (mymattcarroll) [inline DOCUMENTATION.md:74]:
```suggestion
- For _Forms List_ type apps, the form can be in the _Forms List_ menu item.
```

File: DOCUMENTATION.md [hunk 8]
```diff
@@ -51,8 +51,27 @@ Various Types used in this package can also be imported:
...
+
+Actions that require a developer key to be assigned to an app will have the following in the function definition:
+
+**App Association Required**
+
+These actions require the developer key to be associated with an app that is associated with the form that is being actioned. A form can be associated with an app in the following ways:
+
+- For _Forms List_ type apps, the form must be in the Forms List menu item
+- For _Tiles_ type apps, the form can be in a _Container_ menu item or added directly to the menu via a _Form_ menu item.
+- For _Tiles_ or _Forms List_ type apps, the form can be assigned to an action in a scheduled task.
```
Comment (Zaxist) [inline DOCUMENTATION.md:76]:
```suggestion
- For _Tiles_ or _Forms List_ type apps, the form can be assigned to an action in a _Scheduled task_.
```

File: DOCUMENTATION.md [hunk 9]
```diff
@@ -51,8 +51,27 @@ Various Types used in this package can also be imported:
...
+- For _Approvals_ type apps, the form can be an approval form.
```
Comment (Zaxist) [inline DOCUMENTATION.md:77]:
```suggestion
- For _Approvals_ type apps, the form can be an _Approval form_.
```

File: DOCUMENTATION.md (hunk 4 above)
Comment (Zaxist) [reply DOCUMENTATION.md:72]: Yeh but that might not always be the case. And tying the two concepts together is more for the user to understand IMO.

Comment (Zaxist) [reply DOCUMENTATION.md:72]: I'm looking at it this way: Currently all the actions that require an app association just happen to be single form actions.

**Pull request merged**